infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogCompaction.java

```diff
@@ -0,0 +1,54 @@
+package org.corfudb.infrastructure.log;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.util.MetricsUtils;
+
+import java.util.concurrent.*;
+
+/**
+ * Scheduled log compaction manager.
+ * Execute log compaction safely (prevent scheduled executor corruption if the task throws an exception)
+ * https://stackoverflow.com/questions/6894595/scheduledexecutorservice-exception-handling
+ */
+@Slf4j
+public class StreamLogCompaction {
+    public static final String STREAM_COMPACT_METRIC = StreamLogCompaction.class.getName();
+
+    private final ThreadFactory threadFactory = new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setNameFormat("LogUnit-Maintenance-%d")
+            .build();
+
+    /**
+     * How many times log compaction executed
+     */
+    private final Counter gc = CorfuRuntime.getDefaultMetrics().counter(STREAM_COMPACT_METRIC);
+
+    /**
+     * A scheduler, which is used to schedule periodic garbage collection.
+     */
+    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(threadFactory);
+    private final ScheduledFuture<?> compactor;
+
+    public StreamLogCompaction(StreamLog streamLog, long initialDelay, long period, TimeUnit timeUnit) {
+        Runnable task = () -> {
+            gc.inc();
+            log.debug("Start log compaction. GC counter: {}", gc.getCount());
+            try {
+                streamLog.compact();
+            } catch (Throwable ex) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)